Bounded mode polymorphism

EliasC opened the pull request:
This commit adds support for adding bounds on the mode of a type
parameter. It also introduces the mode `sharable` which abstracts
over all modes that may be safely shared between active
objects (`read`, `active` and `shared`). This mode can only be
used for type parameters for now, where it allows a polymorphic
context to pass values between active objects. See the changed/added
tests for more details.

EliasC committed:
Support for bounded mode polymorphism

This commit adds support for adding bounds on the mode of a type
parameter. It also introduces the mode `sharable` which abstracts
over all modes that may be safely shared between active
objects (`read`, `active` and `shared`). This mode can only be
used for type parameters for now, where it allows a polymorphic
context to pass values between active objects.

EliasC committed:
Refactoring

supercooldave commented on src/types/Typechecker/TypeError.hs:
```diff
@@ -763,6 +775,16 @@ instance Show Error where
                 "be called by a different active object")
                (show name)
+    show (PolymorphicArgumentSendError arg ty) =
+        printf ("Cannot pass polymorphic expression '%s' between active " ++
```
"expression" might not be the right word. You cannot pass an expression.

EliasC replied:
"Cannot pass polymorphic value"?

supercooldave replied:
Or more verbosely, "values of polymorphic type".

EliasC replied:
> Cannot pass value of '%s' between active objects. Its type is polymorphic so it may not be safe to share.

supercooldave replied:
Looks good.

supercooldave commented on src/types/Typechecker/TypeError.hs:
```diff
@@ -799,9 +821,17 @@ instance Show Error where
-               (show ty)
+    show (UnsafeTypeArgumentError formal ty) =
+        if isModeless ty then
+          printf "Cannot use non-aliasable type '%s' as type argument" --TODO: This seems imprecise...
```
Is non-aliasable the same as linear?

EliasC replied:
The `TODO` there is relevant... I'll look into when this case is hit exactly and change it accordingly. It might be that this is actually dead code.

supercooldave replied:
Let me know.

EliasC replied:
Actually, the code here makes sense, albeit it's not very precise. This branch is for when we have a type that is indirectly `linear`, for example a tuple where one element is `linear`. It doesn't say "linear" in the error message as it gives the same error for `subord`inate types. I've kept the TODO, but elaborated on it a bit to make it more clear how the error message might be improved in the future.

supercooldave reviewed (commented):
Minor comments for now.

supercooldave commented on modules/standard/Active/ActiveTest.enc:
```diff
@@ -1,12 +1,13 @@
--- This file was automatically converted by encorec
-
 import Active.Active
```
Shouldn't this test be moved to the test directory?
(Yes, you can argue that you didn't put it here, but as you are modifying the test, then now is the time!!)

EliasC replied:
I would even argue that it *is* my responsibility, as it's PR that makes this library (verifiably) safe to use again.

supercooldave replied:
Let me know.

EliasC replied:
Moved!

supercooldave replied:
Has it moved?

EliasC replied:
Hm... Copied seems more correct. Removing the old one now!

supercooldave commented:
I'm pretty happy with this. My feeling is that we should merge it, after the very minor comments are addressed.
Does anyone else want to look at it?

EliasC committed:
Reacted to @supercooldave's comments

EliasC commented:
I've reacted to Dave's comments.

EliasC committed:
Only allow modes for type parameters

Changed parsing so that type variables can only have modes in the
parameter list of functions/classes/etc.

EliasC committed:
Remove old test

EliasC commented:
Pushed additional changes:
* Changed the parser so that it only allows modes for type variables when they appear in type parameter lists
* Actually moved `ActiveTest.enc`
